### DISCUSSION TITLE
Manage Trusted Extensions for custom AuthenticationProvider

### DISCUSSION THREAD

**Discussion by AnnaSalata** [microsoft/vscode-discussions#1583] [Extension Development QnA] [answered]
I'm currently creating an authentication provider for my extension and trying to figure out, how to deal with the Manage Trusted Extensions VS Code functionality for authentication provider under VS Code Accounts section. 

When removing extension from the list of trusted extensions, _Grant access to {Auth Provider} for {Extension}... (1)_ appears and doesn't disappear when check it back. The behavior is different, using two ways of granting access back (please see the description below). Is there any possibility to determine, if my extension trusted or not? I haven't found any API for this aim.
Has anyone any ideas on how to deal with this issue?

### Unchecking extension, using Manage Trusted Extensions

1. When user is signed in, there's an option to _Manage Trusted Extensions_:
<img width="547" alt="image" src="https://github.com/user-attachments/assets/c155e961-6912-49f0-8c0f-5b414bbf1522">

2. When this option is clicked, a list of trusted extensions appears and there's a possibility to uncheck my extension:
<img width="626" alt="image" src="https://github.com/user-attachments/assets/44b5776f-6e9f-434d-974c-1fbb7a074135">

3. When user unchecks it and presses OK
<img width="626" alt="image" src="https://github.com/user-attachments/assets/23f98eda-f22f-4fc5-acb1-7ed9cd4e5de2">

4. An event is fired inside my **vscode.authentication.onDidChangeSessions** subscription and session becomes unavailable:
<img width="526" alt="image" src="https://github.com/user-attachments/assets/baa226c3-7796-4611-8596-ac2cd22ee7f8">

5. At the same time an option "_Grant access to My Provider for My Extension... (1)_" appears under the Accounts section 
<img width="421" alt="image" src="https://github.com/user-attachments/assets/60a71c21-b9de-47a1-9d8c-51476938cf28">

### Granting access, case 1. Using "Grant access to My Provider for My Extension... (1)" option

1. If user presses "_Grant access to My Provider for My Extension... (1)_", a confirmation popup appears
<img width="262" alt="image" src="https://github.com/user-attachments/assets/ddb29194-98e2-4219-ab75-2fe97454ef7e">

2. And after pressing **Allow**, an event is fired inside my **vscode.authentication.onDidChangeSessions** subscription and session becomes available:
<img width="524" alt="image" src="https://github.com/user-attachments/assets/d24acc06-d44a-4a13-b699-e5ee6cb30faa">

3. "_Grant access to My Provider for My Extension... (1)_" disappears from Accounts section:
<img width="377" alt="image" src="https://github.com/user-attachments/assets/05ae3d40-2ad1-4ef6-baff-ea9c56b9489c">

### Granting access, case 2. Checking extension as trusted, using account -> Manage Trusted Extensions

1. If user ignores the possibility to "_Grant access to My Provider for My Extension... (1)_" under the Accounts section and again goes to the account -> _Manage Trusted Extensions_ 
<img width="604" alt="image" src="https://github.com/user-attachments/assets/46d49432-2691-4f68-aa0c-ddaba044ac57">

2. Checks the extension as trusted and confirms this by pressing OK
<img width="609" alt="image" src="https://github.com/user-attachments/assets/06cf8611-3325-44b0-b3c7-bd1673b181e9">

3. An event is fired, session becomes available:
<img width="526" alt="image" src="https://github.com/user-attachments/assets/965a0457-4d10-4142-a281-36405b78a805">

4. But "_Grant access to My Provider for My Extension... (1)_" under Accounts section doesn't disappear, even though access is actually granted:
<img width="425" alt="image" src="https://github.com/user-attachments/assets/8d33f630-91d7-45dd-b562-88cb14d1791d">

In my case I need to make checking extension as trusted work the same way as when "_Grant access to My Provider for My Extension... (1)_" is clicked.

**Comment by isidorn** [comment, marked as answer]
This is currently not exposed through API.

**Reply by AnnaSalata** [reply]
Hi, @isidorn
This behavior leads to the further issues, such as possibility to sign in with at least two accounts, even if AuthProvider is configured with { supportsMultipleAccounts: false } flag. I've described all the details in the bug I've opened:
https://github.com/microsoft/vscode/issues/227694
It would be nice to provide corresponding API to have possibility to manage it from the extension, or fix it on the VS Code side, when user grants access to the extension back indirectly, using Manage Trusted Extensions -> checking extension and pressing OK.
